compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -913,121 +414,454 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+        while let Some(inner) = walker.next() {
+            if self.generic_arg_is_target(inner) {
+                return true;
+            }
+            match inner.unpack() {
+                GenericArgKind::Lifetime(_) => {}
+                GenericArgKind::Type(ty) => {
+                    if matches!(ty.kind(), ty::Opaque(..)) {
+                        // Opaque types can't be named by the user right now
+                        // FIXME(type_alias_impl_trait): These opaque types
+                        // can actually be named, so it would make sense to
+                        // adjust this case and add a test for it.
+                        walker.skip_current_subtree();
+                    }
+                }
+                GenericArgKind::Const(ct) => {
+                    if matches!(ct.val, ty::ConstKind::Unevaluated(..)) {
+                        // You can't write the generic arguments for
+                        // unevaluated constants.
+                        walker.skip_current_subtree();
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    fn expr_inferred_subst_iter(
+        &self,
+        expr: &'tcx hir::Expr<'tcx>,
+    ) -> Box<dyn Iterator<Item = InsertableGenericArgs<'tcx>> + 'a> {
+        let tcx = self.infcx.tcx;
+        let typeck_results = self.infcx.in_progress_typeck_results.unwrap().borrow();
+        match expr.kind {
+            hir::ExprKind::Path(ref path) => {
+                if let Some(substs) = typeck_results.node_substs_opt(expr.hir_id) {
+                    return self.path_inferred_subst_iter(expr.hir_id, substs, path);
+                }
+            }
+            hir::ExprKind::Struct(path, _, _) => {
+                if let Some(ty) = self.opt_node_type(expr.hir_id) {
+                    if let ty::Adt(_, substs) = ty.kind() {
+                        return self.path_inferred_subst_iter(expr.hir_id, substs, path);
+                    }
+                }
+            }
+            hir::ExprKind::MethodCall(segment, _, _) if segment.infer_args => {
+                if let Some(def_id) = typeck_results.type_dependent_def_id(expr.hir_id) {
+                    let generics = tcx.generics_of(def_id);
+                    if generics.has_impl_trait() {
+                        return Box::new(iter::empty());
+                    }
+                    let substs = match typeck_results.node_substs_opt(expr.hir_id) {
+                        Some(s) => s,
+                        None => return Box::new(iter::empty()),
+                    };
+                    return Box::new(iter::once(InsertableGenericArgs {
+                        insert_span: segment.ident.span.shrink_to_hi(),
+                        inferred: &substs[substs.len() - generics.params.len()..],
+                        generics_def_id: def_id,
+                        def_id,
+                    }));
                 }
             }
```

Comment:
I think that in these two branches we could recover the logic we previously had to point only at the last relevant method call in the chain instead of the whole expression (as well as just the struct literal name instead of the whole struct literal with its fields) if we make `FindInferSourceVisitor` also keep the span we'll use for the actual error.

Reply:
do you have a test where this is an issue? in `ui/const-generics/infer/method-chain` this already looks about right to me